### PR TITLE
Unordered sequences

### PR DESCRIPTION
**Pull request by jadams-tresys**
Have an initial implementation working with simple test cases. Currently
still having issues with some negative test cases that still work due to
lack of validation. Also still have a circular definition error in
contentStartAlignment, which calls priorSibsAlignmentsApprox which calls
contentStartAlignment.

DAFFODIL-1159

### CODE REVIEW

**Comment by jadams-tresys** [comment]
I h ave pushed up another commit to this that finishes the implementation of Unordered Sequences, with the exception of one test (test_simple_nil) that seems to cause an infinite loop for some reason.  The unordered sequences in the vCard also work now, athough there are a few other things to clean up in that schema before it is working 100%.

File: daffodil-core/src/main/scala/org/apache/daffodil/grammar/AlignedMixin.scala [hunk 1]
```diff
@@ -260,6 +285,7 @@ trait AlignedMixin extends GrammarMixin { self: Term =>
           case LengthKind.Prefixed => LengthMultipleOf(1) // NYI
         }
       }
+      case mg: ModelGroup => LengthMultipleOf(1)
```
[the remark this answers is not in the document]
Comment (jadams-tresys) [reply daffodil-core/src/main/scala/org/apache/daffodil/grammar/AlignedMixin.scala]: So, this code path is actually getting hit by a negative test that is trying to trigger an SDE.  The test has an unordered sequence containing a choice, which is not allowed.  So what I think is happening is that the we are attempting to get the elementSpecifiedLengthApprox of this choice.
I think the proper fix for this is to use the Assert.usageError for now, and handle this case in the SequenceGrammarMixin so that it doesn't even try to make a SequenceChild from the ModelGroup in an unordered sequence.

File: daffodil-core/src/main/scala/org/apache/daffodil/grammar/AlignedMixin.scala [hunk 2]
```diff
@@ -52,10 +52,20 @@ trait AlignedMixin extends GrammarMixin { self: Term =>
...
       true
+    } else if (nearestEnclosingSequence.isDefined && !nearestEnclosingSequence.get.isOrdered) {
+      // Enclosed within an unordered sequence, all items in the seq
+      // must have a length that will leave the sequence aligned for
+      // isKnownToBeAligned to be true
+      val (priorSibs, optEnclosingParent) = potentialPriorTerms
+      priorSibs.forall(x => {
+        val contentAlignment = AlignmentMultipleOf((x.elementSpecifiedLengthApprox + x.trailingSkipApprox).nBits)
+        (contentAlignment % alignmentApprox) == 0
+      })
```
[the remark this answers is not in the document]
Comment (jadams-tresys) [reply daffodil-core/src/main/scala/org/apache/daffodil/grammar/AlignedMixin.scala]: I have made changes to make this reflect the self alignment code for arrays, which prevents the circularity errors I was getting.

File: daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/parsers/SequenceParserBases.scala [hunk 3]
```diff
@@ -44,6 +49,23 @@ abstract class SequenceParserBase(
...
+  childParsersArg: Vector[Parser])
+  extends SequenceParserBase(srd, childParsersArg) {
+
+  override protected def parse(pstate: PState) = {
+
+    do {
+      choiceParser.parse(pstate)
+    } while (pstate.processorStatus eq Success)
+
+    pstate.setSuccess()
```
[the remark this answers is not in the document]
Comment (jadams-tresys) [reply daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/parsers/SequenceParserBases.scala]: I have added validation code for min/max occurs, although it ended up being more complex than originally expected.  Given that arrays within unordered sequences do not need to be contiguous, we were creating multiple DIArrays for the same element.  I was originally going to prevent the creation of multiple arrays, but that had issues if we backtracked.  So, in order to get a proper count we had to combine the arrays.

File: daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/parsers/SequenceParserBases.scala [hunk 4]
```diff
@@ -44,6 +49,23 @@ abstract class SequenceParserBase(
...
+  srd: SequenceRuntimeData,
+  choiceParser: ChoiceParser,
+  childParsersArg: Vector[Parser])
+  extends SequenceParserBase(srd, childParsersArg) {
+
+  override protected def parse(pstate: PState) = {
+
+    do {
+      choiceParser.parse(pstate)
+    } while (pstate.processorStatus eq Success)
```
[the remark this answers is not in the document]
Comment (jadams-tresys) [reply daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/parsers/SequenceParserBases.scala]: While I am not 100%, I think most of the complicated separator logic is handled by the separator parser.  The OrderedSequnceParser also just prepends the separator parser to the list of sequence child parsers.
From what I can tell, the complicated logic in the ordered parser simply isn't applicable to the unordered stuff as elements may not be in order and arrays may not be contiguous.  I've added several more tests to verify that separators are handled correctly and everything seems to be working correctly.

File: daffodil-core/src/main/scala/org/apache/daffodil/grammar/primitives/SequenceCombinator.scala [hunk 5]
```diff
@@ -76,3 +78,63 @@ class OrderedSequence(sq: SequenceTermBase, sequenceChildrenArg: Seq[SequenceChi
...
+
+  private lazy val sequenceChildren = sequenceChildrenArg.toVector
+
+
+  private lazy val parsers = {
+    val res = alternatives.map { alt =>
+      val p = alt.parser
+      val res =
+        if (p.isEmpty)
+          new EmptyChoiceBranchParser(alt.context.runtimeData)
```
[the remark this answers is not in the document]
Comment (mbeckerle) [reply daffodil-core/src/main/scala/org/apache/daffodil/grammar/primitives/SequenceCombinator.scala]: Steve is correct. All children of an unordered sequence must be elements, so this case is not possible.

File: daffodil-core/src/main/scala/org/apache/daffodil/grammar/primitives/SequenceCombinator.scala [hunk 6]
```diff
@@ -76,3 +78,63 @@ class OrderedSequence(sq: SequenceTermBase, sequenceChildrenArg: Seq[SequenceChi
...
+          p
+      res
+    }
+    res
+  }
+
+
+  override lazy val parser: Parser = sq.hasSeparator match {
+    case true => {
+      lazy val choiceParser = new ChoiceParser(srd, sepParser +: parsers.toVector)
```
[the remark this answers is not in the document]
Comment (mbeckerle) [reply daffodil-core/src/main/scala/org/apache/daffodil/grammar/primitives/SequenceCombinator.scala]: I agree with steve that this isn't correct. The choice parser should not include the separator.  because below you are passing the sepParser and the choiceParser as two different args to UnorderedSeparatedSequenceParser. That should orchestrate finding the sep with the sepParser when it is expected. 

Tests that would trip this up would be using dfdl:separatorPosition="postfix" I think, because with the sep parser as part of the choice parser, you can't ever have a final sep without a choice child.

File: daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/parsers/SeparatedSequenceParsers.scala [hunk 7]
```diff
@@ -123,3 +123,13 @@ final class OrderedSeparatedSequenceParser(
...
 }
+
+final class UnorderedSeparatedSequenceParser(
+  rd: SequenceRuntimeData,
+  spos: SeparatorPosition,
+  sep: Parser,
+  choiceParser: ChoiceParser)
+  extends UnorderedSequenceParserBase(rd, choiceParser) {
+
+  override lazy val childProcessors = Vector(choiceParser)
```
Comment (mbeckerle) [inline daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/parsers/SeparatedSequenceParsers.scala]: Right here, above is where you want to concatenate the sep parser with the choice parser. This childProcessors member is just there to satisfy a need to know all sub parsers (regardless of their role) have been properly initialized. 

All the rest of the logic needs to distinguish the sep parser from the choiceParser so as to handle when they appear together and separately (e.g., separatorPositon="postfix" case). 

Rather than melding the sep into the choiceParser, and then using your UnorderedSequenceParserBase, you should not meld it, and reuse OrderedSequenceParser, providing it the sep parser, and a single child parser which is the choiceParser. Then you postprocess after that completes to reorg the infoset which has artifacts of being assembled in unordered manner that you need to fixup to have it logically be ordered.

File: daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/parsers/SequenceParserBases.scala [hunk 8]
```diff
@@ -44,6 +50,45 @@ abstract class SequenceParserBase(
...
+abstract class UnorderedSequenceParserBase(
+  srd: SequenceRuntimeData,
+  choiceParser: ChoiceParser)
+  extends SequenceParserBase(srd, Vector(choiceParser)) {
+
+  override protected def parse(pstate: PState) = {
+
+    var isDone = false
+
+    while (!isDone && (pstate.processorStatus eq Success)) {
```
Comment (mbeckerle) [inline daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/parsers/SequenceParserBases.scala]: If this loop is being used in the separated case, then this can't work because it can't handle separatorPosition="postfix". It has no way to decide to parse a choice child followed by a postfix separator because it is assuming the sep parser has been melded into the choiceParser.

Comment (jadams-tresys) [reply daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/parsers/SequenceParserBases.scala]: While I'm not saying that I am doing things correctly, I do have tests with separatorPosition=prefix and postfix that all produce the expected results.  That being said, I'm sure I haven't tested every edge case.

I'll start working on making the changes to use OrderedSeparatedParser instead of the Unordered one.

Comment (jadams-tresys) [reply daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/parsers/SequenceParserBases.scala]: Just talked with Steve and I agree that this is wrong and probably wouldn't work for some negative test cases.  I'll add a few of those to verify that the new code is working correctly.

File: daffodil-runtime1/src/main/scala/org/apache/daffodil/infoset/InfosetImpl.scala [hunk 9]
```diff
@@ -1415,6 +1415,10 @@ sealed class DIComplex(override val erd: ElementRuntimeData, val tunable: Daffod
     }
   }
 
+  final def sortChildNodesByPosition(): Unit = {
+    childNodes = childNodes.sortBy(_.erd.position)
+  }
```
[the remark this answers is not in the document]
Comment (jadams-tresys) [reply daffodil-runtime1/src/main/scala/org/apache/daffodil/infoset/InfosetImpl.scala]: DAFFODIL-2198 has been created to track this issue.

File: daffodil-core/src/main/scala/org/apache/daffodil/grammar/AlignedMixin.scala [hunk 10]
```diff
@@ -47,12 +47,14 @@ case class LengthMultipleOf(nBits: Long) extends LengthApprox
 
 trait AlignedMixin extends GrammarMixin { self: Term =>
 
+  val isInUnorderedSequence: Boolean = nearestEnclosingSequence.isDefined && !nearestEnclosingSequence.get.isOrdered
+
```
[the remark this answers is not in the document]
Comment (jadams-tresys) [reply daffodil-core/src/main/scala/org/apache/daffodil/grammar/AlignedMixin.scala]: Will do, kind of surprised I didn't already have duplicate calls for something like this.

File: daffodil-core/src/main/scala/org/apache/daffodil/grammar/AlignedMixin.scala [hunk 11]
```diff
@@ -170,18 +172,29 @@ trait AlignedMixin extends GrammarMixin { self: Term =>
...
+        if (isInUnorderedSequence) {
+          Seq(alignmentApprox + (elementSpecifiedLengthApprox + trailingSkipApprox))
+        } else {
+          Seq()
+        }
+
       val priorSibsAlignmentsApprox = priorSibs.map { ps =>
-        val eaa = ps.endingAlignmentApprox
+        val eaa = if (isInUnorderedSequence) {
+          alignmentApprox + (elementSpecifiedLengthApprox + trailingSkipApprox)
```
[the remark this answers is not in the document]
Comment (jadams-tresys) [reply daffodil-core/src/main/scala/org/apache/daffodil/grammar/AlignedMixin.scala]: I can't say I fully understand the alignment code, but I think this is OK, I just forgot to call **ps**.elementSpecifiedLengthApprox + **ps.**trailingSkipApprox, which works as expected.

File: daffodil-runtime1/src/main/scala/org/apache/daffodil/infoset/InfosetImpl.scala [hunk 12]
```diff
@@ -1007,6 +1006,15 @@ final class DIArray(
     ie.setArray(this)
   }
 
+  def concat(array: DIArray): DIArray = {
+    val newContents = array.contents
+    newContents.map( ie => {
```
[the remark this answers is not in the document]
Comment (jadams-tresys) [reply daffodil-runtime1/src/main/scala/org/apache/daffodil/infoset/InfosetImpl.scala]: Made the change to not return anything.  Could no longer use reduceLeft in flatten and had to switch to a less pretty foreach.

File: daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/parsers/SequenceParserBases.scala [hunk 13]
```diff
@@ -30,24 +36,58 @@ import org.apache.daffodil.processors.TermRuntimeData
...
+      erds.get.foreach(e => {
+        val min = e.minOccurs
+        val max = e.maxOccurs
+        val isUnbounded = max == -1
+
+        if (isUnbounded)
+          pstate.validationError("Element %s failed check of minOccurs='%s' and maxOccurs='UNBOUNDED", e.diagnosticDebugName, min)
+        else
+          pstate.validationError("Element %s failed check of minOccurs='%s' and maxOccurs='%s'", e.diagnosticDebugName, min, max)
+      })
```
[the remark this answers is not in the document]
Comment (jadams-tresys) [reply daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/parsers/SequenceParserBases.scala]: You are right, for some reason I thought I needed this logic outside of the infoset, but I don't think that is the case.  This should also simplify flattendAndValidate code too, instead of having to maintain a list to return.

[the remark this answers is not in the document]
Comment (jadams-tresys) [reply daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/parsers/SequenceParserBases.scala]: Actually, is there a way to get the PState object from the infoset? I don't see anything obvious, which I think might be why I did the actual validation error in SequenceParserBases

[the remark this answers is not in the document]
Comment (stevedlawrence) [reply daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/parsers/SequenceParserBases.scala]: I guess you coul pass the pstate into this function. Not sure if we do that anywhere is the infoset code, but it doesn't seem totally unreasonable.

[the remark this answers is not in the document]
Comment (stevedlawrence) [reply daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/parsers/SequenceParserBases.scala]: Actually, you could make the parameter a ThrowsSDE, and then pass the pstate into that. That way if other things needed to use that that had a ThrowsSDE context but not a PState it coudl still be used. Makes it less tied to parsing. I *think* ThrowsSDE also can also create validation errors?

File: daffodil-runtime1/src/main/scala/org/apache/daffodil/infoset/InfosetImpl.scala [hunk 14]
```diff
@@ -1415,6 +1423,50 @@ sealed class DIComplex(override val erd: ElementRuntimeData, val tunable: Daffod
...
+    nodes.foreach( childNodes ++= _ )
+    if (ret.isEmpty)
+      Maybe.Nope
+    else
+      Maybe(ret)
+  }
+
   override def addChild(e: InfosetElement): Unit = {
     if (!e.isHidden && !hasVisibleChildren) hasVisibleChildren = true
     if (e.runtimeData.isArray) {
```
[the remark this answers is not in the document]
Comment (jadams-tresys) [reply daffodil-runtime1/src/main/scala/org/apache/daffodil/infoset/InfosetImpl.scala]: So instead of actually flattening this, I updated the addChildToFastLookup to concat DIArrays, which resolves this issue and doesn't have any issues with backtracking.  I think there is already appropriate cleanup code in the restoreFrom function.

I have also added a test to verify this.

File: daffodil-runtime1/src/main/scala/org/apache/daffodil/infoset/InfosetImpl.scala [hunk 15]
```diff
@@ -1415,6 +1423,50 @@ sealed class DIComplex(override val erd: ElementRuntimeData, val tunable: Daffod
     }
   }
 
+  final def sortChildNodesByPosition(): Unit = {
+    // TODO: Once we upgrade to Scala 2.13 we can call the inplace sort method: DAFFODIL-2152
+    // childNodes = childNodes.sortInPlace(_.erd.position)
```
[the remark this answers is not in the document]
Comment (jadams-tresys) [reply daffodil-runtime1/src/main/scala/org/apache/daffodil/infoset/InfosetImpl.scala]: I think this is fine, as it only gets called by the unordered parser, which means it will only sort the children in the unordered sequence.  In the tests I have, including tests where the unordered sequence is part of a larger sequence, childNodes only contains the elements of the unordered sequence.

[the remark this answers is not in the document]
Comment (stevedlawrence) [reply daffodil-runtime1/src/main/scala/org/apache/daffodil/infoset/InfosetImpl.scala]: Are you sure? If you have something like:
```xml
<xs:element name="root">
  <xs:complexType>
    <xs:sequence>
      <xs:element name="a" />
      <xs:element name="b" />
      <xs:sequence dfdl:sequenceKind="unordered">
        <xs:element name="c" />
        <xs:element name="d" />
     </xs:sequence>
  </xs:complexType>
</xs:element>
```
In that case elements a, b, c, and d are all children of the root element, and I would think would all be in the same childNodes array.

File: daffodil-runtime1/src/main/scala/org/apache/daffodil/infoset/InfosetImpl.scala (hunk 14 above)
[the remark this answers is not in the document]
Comment (stevedlawrence) [reply daffodil-runtime1/src/main/scala/org/apache/daffodil/infoset/InfosetImpl.scala]: I feel like there's an edge case where that wouldn't work. Imagine we parse an array, and then parse a scalar. Now say we try to parse another of the same array. WIth your change that will contact to the previous array. But then if that array element fails for some reason we need to be able to remove that array we concated. But I *think* the way we backtrack arrays is by looking the last element in the infoset and reducing the array size. But the array is not the last in the childNodes seq now, since the array we added is some previous array.

File: daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/parsers/SequenceParserBases.scala [hunk 16]
```diff
@@ -224,12 +227,22 @@ abstract class OrderedSequenceParserBase(
               pstate.setSuccess()
               isDone = true
             }
+            case (MissingItem | MissingSeparator | FailureUnspecified) if (!isOrdered) => {
+              pstate.setSuccess()
```
Comment (mbeckerle) [inline daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/parsers/SequenceParserBases.scala]: Add a comment here as to why for unordered these failure conditions are ok, which I think is because in general unordered stuff behaves like occursCountKind parsed, i.e., it parses until something stops it by a failure.

File: daffodil-test/src/test/resources/org/apache/daffodil/section14/unordered_sequences/UnorderedSequences.tdml [hunk 17]
```diff
@@ -869,10 +758,123 @@
...
+					<e1>1</e1>
+					<e2>2</e2>
+					<e3>3</e3>
+				</infix>
+			</tdml:dfdlInfoset>
+		</tdml:infoset>
+  </tdml:parserTestCase>
+
+	<tdml:parserTestCase name="test_separated_prefix"
+		model="separated" description="Section 14 Sequence group with left over data - DFDL-14-001R"
```
Comment (mbeckerle) [inline daffodil-test/src/test/resources/org/apache/daffodil/section14/unordered_sequences/UnorderedSequences.tdml]: Are these test descriptions just left over from cut-paste. I don't think they match what you are testing here. I would just delete them. Your test names are descriptive enough.

Comment (jadams-tresys) [reply daffodil-test/src/test/resources/org/apache/daffodil/section14/unordered_sequences/UnorderedSequences.tdml]: Will do.

File: daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/parsers/SequenceParserBases.scala (hunk 16 above)
Comment (jadams-tresys) [reply daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/parsers/SequenceParserBases.scala]: Will do.

File: daffodil-runtime1/src/main/scala/org/apache/daffodil/infoset/InfosetImpl.scala [hunk 18]
```diff
@@ -1450,37 +1449,40 @@ sealed class DIComplex(override val erd: ElementRuntimeData, val tunable: Daffod
...
-        val isUnbounded = max0 == -1
+        val min = erd.minOccurs
+        val max = erd.maxOccurs
+        val isUnbounded = max == -1
 
         // Flatten multiple DIArrays into the first one
         if (erd.isArray) {
-          nodes.reduceLeft((a, b) => a.asInstanceOf[DIArray].concat(b.asInstanceOf[DIArray]))
+          val a = nodes(0).asInstanceOf[DIArray]
+          nodes.drop(1).foreach( b => a.concat(b.asInstanceOf[DIArray]))
```
Comment (stevedlawrence) [inline daffodil-runtime1/src/main/scala/org/apache/daffodil/infoset/InfosetImpl.scala]: I think ``.tail`` instead of ``.drop(1)`` is maybe more intuitive?

File: daffodil-runtime1/src/main/scala/org/apache/daffodil/infoset/InfosetImpl.scala [hunk 19]
```diff
@@ -1450,37 +1449,40 @@ sealed class DIComplex(override val erd: ElementRuntimeData, val tunable: Daffod
...
           nodes.reduceToSize(1)
+
+          // Need to also remove duplicates from fastLookup
+          val fastSeq = nameToChildNodeLookup.get(a.namedQName)
+          if (fastSeq != null)
+            fastSeq.reduceToSize(1)
+        } else {
+          if (nodes.length > 1)
+            pstate.validationError("Multiple elements detected for scalar element %s", erd.namedQName)
```
Comment (stevedlawrence) [inline daffodil-runtime1/src/main/scala/org/apache/daffodil/infoset/InfosetImpl.scala]: Maybe it would be helpful to mention that this is in an unordered sequence?

Also, the spec states this should be a processing error, not a validation error. We should have tests to make sure the PE is handled correctly.

File: daffodil-runtime1/src/main/scala/org/apache/daffodil/infoset/InfosetImpl.scala [hunk 20]
```diff
@@ -1450,37 +1449,40 @@ sealed class DIComplex(override val erd: ElementRuntimeData, val tunable: Daffod
...
 
         val occurrence = if (erd.isArray) nodes(0).contents.length else nodes.length
-        if (isUnbounded && occurrence < min0)
-          ret = ret :+ erd
-        else if (!isUnbounded && (occurrence < min0 || occurrence > max0))
-          ret = ret :+ erd
+        if (isUnbounded && occurrence < min)
+          pstate.validationError("Element %s failed check of minOccurs='%s' and maxOccurs='unbounded', actual number of occurrences: %s", erd.namedQName, min, occurrence)
+        else if (!isUnbounded && (occurrence < min || occurrence > max))
+          pstate.validationError("Element %s failed check of minOccurs='%s' and maxOccurs='%s', actual number of occurrences: %s", erd.namedQName, min, max, occurrence)
```
Comment (stevedlawrence) [inline daffodil-runtime1/src/main/scala/org/apache/daffodil/infoset/InfosetImpl.scala]: Should these only be one if ``isArray == true``? Since the else above requires that scalar elements have a length of 1 or else it's a processing error? Otherwise, for scalar elements where more than one is parsed, we could get both a PE and validation errors?

File: daffodil-runtime1/src/main/scala/org/apache/daffodil/infoset/InfosetImpl.scala [hunk 21]
```diff
@@ -1438,20 +1438,24 @@ sealed class DIComplex(override val erd: ElementRuntimeData, val tunable: Daffod
...
   }
 
-  final def sortChildNodesByPosition(): Unit = {
+  final def sortChildNodesByPosition(start: Int): Unit = {
     // TODO: Once we upgrade to Scala 2.13 we can call the inplace sort method: DAFFODIL-2152
     // childNodes.sortInPlace(_.erd.position)
-    childNodes = childNodes.sortBy(_.erd.position)
+    val (a, b) = childNodes.splitAt(start)
+    childNodes.clear()
+    childNodes ++= (a ++ b.sortBy(_.erd.position))
```
Comment (stevedlawrence) [inline daffodil-runtime1/src/main/scala/org/apache/daffodil/infoset/InfosetImpl.scala]: Thougts on moving this logic into ``flattendAndValidateChildNodes``? That way we avoid doubling the splitAt, clear, and append work? Essentially turns this function into an unordered sequence post process thing that handles everything about unordered sequences. Could be something like:

```scala
    val (ordered, unordered) = childNodes.splitAt(start)
    childNodes.clear()
    childNodes ++= ordered
    val groups = unordered.groupBy(_.erd)
    val flattenedNodes = groups.map { case (erd, nodes) =>
      // existing array/scala logic here, resulting in a single node
      node
    }
   val sorted = newNodes.sortBy(_.erd.position)
   childNodes ++= sorted
}
```
